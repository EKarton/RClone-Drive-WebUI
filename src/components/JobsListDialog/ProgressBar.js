import Box from '@mui/material/Box';
import LinearProgress from '@mui/material/LinearProgress';
import Typography from '@mui/material/Typography';
import PropTypes from 'prop-types';
import './ProgressBar.scss';

export default function ProgressBar({ numSuccessful, numOngoing }) {
  const percentage = (numSuccessful / (numOngoing + numSuccessful)) * 100;

  return (
    <Box className="progress-bar">
      <Box className="progress-bar__bar-wrapper">
        <LinearProgress variant="determinate" value={percentage} />
      </Box>
      <Box>
        <Typography variant="body2" color="text.secondary" className="progress-bar__text">
          {Math.round(percentage)} %
        </Typography>
      </Box>
    </Box>
  );
}

ProgressBar.propTypes = {
  numSuccessful: PropTypes.number,
  numOngoing: PropTypes.number,
};
