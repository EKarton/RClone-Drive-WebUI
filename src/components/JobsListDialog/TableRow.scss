.table-row {
  &__file-type-cell {
    width: 2rem;
  }

  &__dir-path-cell,
  &__file-name-cell {
    white-space: nowrap;
    overflow: hidden;
    text-overflow: ellipsis;
    padding-right: 1rem;
  }

  &__file-size-cell {
    white-space: nowrap;
    width: 5rem;
  }

  &__status-cell {
    width: 5rem;
  }

  &__cancel-button-cell {
    width: 3rem;
  }
}
