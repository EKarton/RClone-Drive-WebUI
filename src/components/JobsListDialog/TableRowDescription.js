import { JobStatus } from 'services/RCloneJobTracker/constants';
import { getFullPath } from 'utils/filename-utils';

const getFileString = (file) => {
  return getFullRemotePath(file.remote, getFullPath(file.dirPath, file.name));
};

const getFullRemotePath = (remote, fullPath) => {
  return `${remote}:${fullPath}`;
};

function Upload({ status, job }) {
  if (status === JobStatus.ERROR) {
    return (
      <span>
        Failed to upload file {job.name} to {job.remote}:{job.dirPath}
      </span>
    );
  }

  if (status === JobStatus.SUCCESS) {
    return (
      <span>
        Uploaded file {job.name} to {job.remote}:{job.dirPath}
      </span>
    );
  }

  return (
    <span>
      Uploading file {job.name} to {job.remote}:{job.dirPath}
    </span>
  );
}

function MoveFile({ status, job }) {
  if (status === JobStatus.ERROR) {
    return (
      <span>
        Failed to move file from {getFileString(job.src)} to {getFileString(job.target)}
      </span>
    );
  }

  if (status === JobStatus.SUCCESS) {
    return (
      <span>
        Moved file from {getFileString(job.src)} to {getFileString(job.target)}
      </span>
    );
  }

  return (
    <span>
      Moving file from {getFileString(job.src)} to {getFileString(job.target)}
    </span>
  );
}

function MoveFolder({ status, job }) {
  if (status === JobStatus.ERROR) {
    return (
      <span>
        Failed to move folder from {getFileString(job.src)} to {getFileString(job.target)}
      </span>
    );
  }

  if (status === JobStatus.SUCCESS) {
    return (
      <span>
        Moved folder from {getFileString(job.src)} to {getFileString(job.target)}
      </span>
    );
  }

  return (
    <span>
      Moving folder from {getFileString(job.src)} to {getFileString(job.target)}
    </span>
  );
}

function RenameFile({ status, job }) {
  if (status === JobStatus.ERROR) {
    return (
      <span>
        Failed to rename file {job.src.name} to {job.target.name} in {job.src.remote}:
        {job.src.dirPath}
      </span>
    );
  }

  if (status === JobStatus.SUCCESS) {
    return (
      <span>
        Renamed file {job.src.name} to {job.target.name} in {job.src.remote}:
        {job.src.dirPath}
      </span>
    );
  }

  return (
    <span>
      Renaming file {job.src.name} to {job.target.name} in {job.src.remote}:
      {job.src.dirPath}
    </span>
  );
}

function RenameFolder({ status, job }) {
  if (status === JobStatus.ERROR) {
    return (
      <span>
        Failed to rename folder {job.src.name} to {job.target.name} in {job.src.remote}:
        {job.src.dirPath}
      </span>
    );
  }

  if (status === JobStatus.SUCCESS) {
    return (
      <span>
        Renamed folder {job.src.name} to {job.target.name} in {job.src.remote}:
        {job.src.dirPath}
      </span>
    );
  }

  return (
    <span>
      Renaming folder {job.src.name} to {job.target.name} in {job.src.remote}:
      {job.src.dirPath}
    </span>
  );
}

function GenericJob({ status }) {
  if (status === JobStatus.ERROR) {
    return <span>Job failed</span>;
  }

  if (status === JobStatus.SUCCESS) {
    return <span>Job succeeded</span>;
  }

  return <span>Job ongoing</span>;
}

export default function TableRowDescription({ status, job }) {
  if (job.jobType === 'UPLOAD_FILE') {
    return <Upload status={status} job={job} />;
  }

  if (job.jobType === 'MOVE_FILE') {
    return <MoveFile status={status} job={job} />;
  }

  if (job.jobType === 'MOVE_FOLDER') {
    return <MoveFolder status={status} job={job} />;
  }

  if (job.jobType === 'RENAME_FILE') {
    return <RenameFile status={status} job={job} />;
  }

  if (job.jobType === 'RENAME_FOLDER') {
    return <RenameFolder status={status} job={job} />;
  }

  return <GenericJob status={status} />;
}
