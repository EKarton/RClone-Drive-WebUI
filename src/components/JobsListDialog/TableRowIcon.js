import BackupIcon from '@mui/icons-material/Backup';
import DriveFileMoveIcon from '@mui/icons-material/DriveFileMove';
import DriveFileRenameOutlineIcon from '@mui/icons-material/DriveFileRenameOutline';
import TaskIcon from '@mui/icons-material/Task';
import Tooltip from '@mui/material/Tooltip';
import PropTypes from 'prop-types';
import { JobTypes } from 'utils/constants';

export default function TableRowIcon({ jobType }) {
  const renderIcon = () => {
    switch (jobType) {
      case JobTypes.UPLOAD_FILE: {
        return <BackupIcon />;
      }
      case JobTypes.MOVE_FILE:
      case JobTypes.MOVE_FOLDER: {
        return <DriveFileMoveIcon />;
      }
      case JobTypes.RENAME_FILE:
      case JobTypes.RENAME_FOLDER: {
        return <DriveFileRenameOutlineIcon />;
      }
      default: {
        return <TaskIcon />;
      }
    }
  };

  return (
    <Tooltip title={jobType} arrow>
      <span>{renderIcon()}</span>
    </Tooltip>
  );
}

TableRowIcon.propTypes = {
  jobType: PropTypes.string,
};
