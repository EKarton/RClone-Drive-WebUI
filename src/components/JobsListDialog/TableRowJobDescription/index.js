import { JobTypes } from 'utils/constants';
import GenericJobDescription from './GenericJobDescription';
import MoveFileJobDescription from './MoveFileJobDescription';
import MoveFolderJobDescription from './MoveFolderJobDescription';
import RenameFileJobDescription from './RenameFileJobDescription';
import RenameFolderJobDescription from './RenameFolderJobDescription';
import UploadJobDescription from './UploadJobDescription';

export default function TableRowDescription({ status, job }) {
  if (job.jobType === JobTypes.UPLOAD_FILE) {
    return <UploadJobDescription status={status} job={job} />;
  }

  if (job.jobType === JobTypes.MOVE_FILE) {
    return <MoveFileJobDescription status={status} job={job} />;
  }

  if (job.jobType === JobTypes.MOVE_FOLDER) {
    return <MoveFolderJobDescription status={status} job={job} />;
  }

  if (job.jobType === JobTypes.RENAME_FILE) {
    return <RenameFileJobDescription status={status} job={job} />;
  }

  if (job.jobType === JobTypes.RENAME_FOLDER) {
    return <RenameFolderJobDescription status={status} job={job} />;
  }

  return <GenericJobDescription status={status} />;
}
