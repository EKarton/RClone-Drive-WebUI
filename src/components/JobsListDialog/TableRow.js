import CancelIcon from '@mui/icons-material/Cancel';
import IconButton from '@mui/material/IconButton';
import TableCell from '@mui/material/TableCell';
import MuiTableRow from '@mui/material/TableRow';
import PropTypes from 'prop-types';
import { useEffect, useState } from 'react';
import { BehaviorSubject } from 'rxjs';
import { JobStatus } from 'utils/constants';
import './TableRow.scss';
import TableRowIcon from './TableRowIcon';
import TableRowDescription from './TableRowJobDescription';

export default function TableRow({ job, ...props }) {
  const [status, setStatus] = useState(null);

  useEffect(() => {
    const subscriber = job.status.subscribe((newStatus) => {
      setStatus(newStatus);
    });

    return () => {
      subscriber.unsubscribe();
    };
  }, [job]);

  const handleCancelButtonClick = () => {
    job.cancelJob();
  };

  return (
    <MuiTableRow sx={{ '&:last-child td, &:last-child th': { border: 0 } }} {...props}>
      <TableCell className="table-row__file-type-cell">
        <TableRowIcon jobType={job.jobType} />
      </TableCell>
      <TableCell classes={{ root: 'table-row__file-name-cell' }}>
        <TableRowDescription status={status} job={job} />
      </TableCell>
      <TableCell align="right" className="table-row__status-cell">
        {status}
      </TableCell>
      <TableCell align="right" className="table-row__cancel-button-cell">
        <IconButton
          aria-label="cancel job"
          onClick={handleCancelButtonClick}
          disabled={status !== JobStatus.ONGOING}
          data-testid="cancel-job-button"
        >
          <CancelIcon />
        </IconButton>
      </TableCell>
    </MuiTableRow>
  );
}

TableRow.propTypes = {
  file: PropTypes.shape({
    remote: PropTypes.string,
    dirPath: PropTypes.string,
    name: PropTypes.string,
    type: PropTypes.string,
    status: PropTypes.instanceOf(BehaviorSubject),
  }),
};
