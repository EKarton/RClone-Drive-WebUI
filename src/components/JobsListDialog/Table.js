import MuiTable from '@mui/material/Table';
import MuiTableBody from '@mui/material/TableBody';
import MuiTableContainer from '@mui/material/TableContainer';
import MuiTablePagination from '@mui/material/TablePagination';
import { useMemo, useState } from 'react';
import TableRow from './TableRow';

export default function Table({ jobs }) {
  const [pageNum, setPaqeNum] = useState(0);
  const [numRowsPerPage, setNumRowsPerPage] = useState(10);

  const handleChangePage = (_e, newPageNum) => {
    setPaqeNum(newPageNum);
  };

  const handleChangeRowsPerPage = (e) => {
    setNumRowsPerPage(e.target.value);
    setPaqeNum(0);
  };

  const rowsToDisplay = useMemo(() => {
    const startIdx = pageNum * numRowsPerPage;
    const endIdx = (pageNum + 1) * numRowsPerPage;

    return jobs.slice(startIdx, endIdx);
  }, [jobs, numRowsPerPage, pageNum]);

  return (
    <>
      <MuiTableContainer sx={{ maxHeight: '60vh' }}>
        <MuiTable
          aria-label="simple table"
          padding="none"
          sx={{ minWidth: 500, tableLayout: 'fixed' }}
        >
          <MuiTableBody>
            {rowsToDisplay.map((job) => (
              <TableRow key={job.jobId} job={job} data-testid={`job-id-${job.jobId}`} />
            ))}
          </MuiTableBody>
        </MuiTable>
      </MuiTableContainer>
      <MuiTablePagination
        rowsPerPageOptions={[10, 25, 100]}
        component="div"
        count={jobs.length}
        rowsPerPage={numRowsPerPage}
        page={pageNum}
        onPageChange={handleChangePage}
        onRowsPerPageChange={handleChangeRowsPerPage}
      />
    </>
  );
}
