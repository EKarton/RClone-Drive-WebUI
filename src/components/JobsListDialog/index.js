import Dialog from '@mui/material/Dialog';
import DialogContent from '@mui/material/DialogContent';
import DialogTitle from '@mui/material/DialogTitle';
import useJobQueueInfo from 'hooks/jobs/useJobQueueInfo';
import ProgressBar from './ProgressBar';
import Table from './Table';

export default function JobsListDialog({ open, onClose }) {
  const { jobs, statusCounts } = useJobQueueInfo();
  const { numSuccessful, numOngoing } = statusCounts;

  return (
    <Dialog open={open} onClose={onClose} maxWidth="lg" data-testid="jobs-list-dialog">
      <DialogTitle>Jobs List</DialogTitle>
      <DialogContent>
        <ProgressBar numSuccessful={numSuccessful} numOngoing={numOngoing} />
        <Table jobs={jobs} />
      </DialogContent>
    </Dialog>
  );
}
