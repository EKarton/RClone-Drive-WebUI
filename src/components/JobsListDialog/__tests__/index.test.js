import { customRender } from 'test-utils/react';
import JobsListDialog from '../index';

describe('JobsListDialog', () => {
  it('should match snapshot given no jobs', () => {
    const { baseElement } = customRender(<JobsListDialog open onClose={jest.fn()} />);

    expect(baseElement).toMatchSnapshot();
  });
});
