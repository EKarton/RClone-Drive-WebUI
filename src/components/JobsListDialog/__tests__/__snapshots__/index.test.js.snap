// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`JobsListDialog should match snapshot given no jobs 1`] = `
<body
  style="padding-right: 1024px; overflow: hidden;"
>
  <div
    aria-hidden="true"
  />
  <div
    class="MuiModal-root MuiDialog-root css-zw3mfo-MuiModal-root-MuiDialog-root"
    data-testid="jobs-list-dialog"
    role="presentation"
  >
    <div
      aria-hidden="true"
      class="MuiBackdrop-root css-yiavyu-MuiBackdrop-root-MuiDialog-backdrop"
      style="opacity: 1; webkit-transition: opacity 225ms cubic-bezier(0.4, 0, 0.2, 1) 0ms; transition: opacity 225ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;"
    />
    <div
      data-test="sentinelStart"
      tabindex="0"
    />
    <div
      class="MuiDialog-container MuiDialog-scrollPaper css-hz1bth-MuiDialog-container"
      role="presentation"
      style="opacity: 1; webkit-transition: opacity 225ms cubic-bezier(0.4, 0, 0.2, 1) 0ms; transition: opacity 225ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;"
      tabindex="-1"
    >
      <div
        class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation24 MuiDialog-paper MuiDialog-paperScrollPaper MuiDialog-paperWidthLg css-1fu2e3p-MuiPaper-root-MuiDialog-paper"
        role="dialog"
      >
        <h2
          class="MuiTypography-root MuiTypography-h6 MuiDialogTitle-root css-bdhsul-MuiTypography-root-MuiDialogTitle-root"
        >
          Jobs List
        </h2>
        <div
          class="MuiDialogContent-root css-ypiqx9-MuiDialogContent-root"
        >
          <div
            class="progress-bar MuiBox-root css-0"
          >
            <div
              class="progress-bar__bar-wrapper MuiBox-root css-0"
            >
              <span
                aria-valuemax="100"
                aria-valuemin="0"
                aria-valuenow="NaN"
                class="MuiLinearProgress-root MuiLinearProgress-colorPrimary MuiLinearProgress-determinate css-eglki6-MuiLinearProgress-root"
                role="progressbar"
              >
                <span
                  class="MuiLinearProgress-bar MuiLinearProgress-barColorPrimary MuiLinearProgress-bar1Determinate css-5xe99f-MuiLinearProgress-bar1"
                  style="transform: translateX(NaN%);"
                />
              </span>
            </div>
            <div
              class="MuiBox-root css-0"
            >
              <p
                class="MuiTypography-root MuiTypography-body2 progress-bar__text css-r40f8v-MuiTypography-root"
              >
                NaN
                 %
              </p>
            </div>
          </div>
          <div
            class="MuiTableContainer-root css-braypr-MuiTableContainer-root"
          >
            <table
              aria-label="simple table"
              class="MuiTable-root css-1i7vgpz-MuiTable-root"
            >
              <tbody
                class="MuiTableBody-root css-apqrd9-MuiTableBody-root"
              />
            </table>
          </div>
          <div
            class="MuiTablePagination-root css-jtlhu6-MuiTablePagination-root"
          >
            <div
              class="MuiToolbar-root MuiToolbar-gutters MuiToolbar-regular MuiTablePagination-toolbar css-11bfvty-MuiToolbar-root-MuiTablePagination-toolbar"
            >
              <div
                class="MuiTablePagination-spacer css-1psng7p-MuiTablePagination-spacer"
              />
              <p
                class="MuiTablePagination-selectLabel css-pdct74-MuiTablePagination-selectLabel"
              >
                Rows per page:
              </p>
              <div
                class="MuiInputBase-root MuiInputBase-colorPrimary css-16c50h-MuiInputBase-root-MuiTablePagination-select"
                variant="standard"
              >
                <div
                  aria-expanded="false"
                  aria-haspopup="listbox"
                  class="MuiTablePagination-select MuiSelect-select MuiSelect-standard MuiInputBase-input css-194a1fa-MuiSelect-select-MuiInputBase-input"
                  role="button"
                  tabindex="0"
                >
                  10
                </div>
                <input
                  aria-hidden="true"
                  class="MuiSelect-nativeInput css-yf8vq0-MuiSelect-nativeInput"
                  tabindex="-1"
                  value="10"
                />
                <svg
                  aria-hidden="true"
                  class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium MuiTablePagination-selectIcon MuiSelect-icon MuiSelect-iconStandard css-pqjvzy-MuiSvgIcon-root-MuiSelect-icon"
                  data-testid="ArrowDropDownIcon"
                  focusable="false"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="M7 10l5 5 5-5z"
                  />
                </svg>
              </div>
              <p
                class="MuiTablePagination-displayedRows css-levciy-MuiTablePagination-displayedRows"
              >
                0â€“0 of 0
              </p>
              <div
                class="MuiTablePagination-actions"
              >
                <button
                  aria-label="Go to previous page"
                  class="MuiButtonBase-root Mui-disabled MuiIconButton-root Mui-disabled MuiIconButton-colorInherit MuiIconButton-sizeMedium css-zylse7-MuiButtonBase-root-MuiIconButton-root"
                  disabled=""
                  tabindex="-1"
                  title="Go to previous page"
                  type="button"
                >
                  <svg
                    aria-hidden="true"
                    class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
                    data-testid="KeyboardArrowLeftIcon"
                    focusable="false"
                    viewBox="0 0 24 24"
                  >
                    <path
                      d="M15.41 16.09l-4.58-4.59 4.58-4.59L14 5.5l-6 6 6 6z"
                    />
                  </svg>
                </button>
                <button
                  aria-label="Go to next page"
                  class="MuiButtonBase-root Mui-disabled MuiIconButton-root Mui-disabled MuiIconButton-colorInherit MuiIconButton-sizeMedium css-zylse7-MuiButtonBase-root-MuiIconButton-root"
                  disabled=""
                  tabindex="-1"
                  title="Go to next page"
                  type="button"
                >
                  <svg
                    aria-hidden="true"
                    class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
                    data-testid="KeyboardArrowRightIcon"
                    focusable="false"
                    viewBox="0 0 24 24"
                  >
                    <path
                      d="M8.59 16.34l4.58-4.59-4.58-4.59L10 5.75l6 6-6 6z"
                    />
                  </svg>
                </button>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
    <div
      data-test="sentinelEnd"
      tabindex="0"
    />
  </div>
</body>
`;
